// @digdir/designsystemet-react/dist/esm/components/Pagination/index.js
'use client';
import { Pagination as Pagination$1 } from './Pagination.js';
import { PaginationButton } from './PaginationButton.js';
import { PaginationItem } from './PaginationItem.js';
import { PaginationList } from './PaginationList.js';

/**
 * Pagination component, used to navigate through a list of items.
 *
 * @example
 * <Pagination>
 *   <Pagination.List>
 *     <Pagination.Item>
 *       <Pagination.Button aria-label='Forrige side'>Forrige</Pagination.Button>
 *     </Pagination.Item>
 *     <Pagination.Item>
 *       <Pagination.Button aria-label='Side 1'>1</Pagination.Button>
 *     </Pagination.Item>
 *     <Pagination.Item>
 *       <Pagination.Button aria-label='Side 2'>2</Pagination.Button>
 *     </Pagination.Item>
 *   </Pagination.List>
 * </Pagination>
 */
const Pagination = Object.assign(Pagination$1, {
    List: PaginationList,
    Item: PaginationItem,
    Button: PaginationButton,
});
Pagination.List.displayName = 'Pagination.List';
Pagination.Item.displayName = 'Pagination.Item';
Pagination.Button.displayName = 'Pagination.Button';

export { Pagination, PaginationButton, PaginationItem, PaginationList };
