// @digdir/designsystemet-react/dist/esm/components/Table/index.js
'use client';
import { Table as Table$1 } from './Table.js';
import { TableBody } from './TableBody.js';
import { TableCell } from './TableCell.js';
import { TableFoot } from './TableFoot.js';
import { TableHead } from './TableHead.js';
import { TableHeaderCell } from './TableHeaderCell.js';
import { TableRow } from './TableRow.js';

/**
 * Table component, used to display tabular data. Renders a native HTML table element.
 *
 * @example
 * <Table>
 *   <Table.Head>
 *     <Table.Row>
 *       <Table.HeaderCell>Name</Table.HeaderCell>
 *       <Table.HeaderCell>Age</Table.HeaderCell>
 *     </Table.Row>
 *   </Table.Head>
 *   <Table.Body>
 *     <Table.Row>
 *       <Table.Cell>John</Table.Cell>
 *       <Table.Cell>25</Table.Cell>
 *     </Table.Row>
 *   </Table.Body>
 *   <Table.Foot>
 *     <Table.Row>
 *       <Table.Cell>Total</Table.Cell>
 *       <Table.Cell>2</Table.Cell>
 *     </Table.Row>
 *   </Table.Foot>
 * </Table>
 */
const Table = Object.assign(Table$1, {
    Head: TableHead,
    Body: TableBody,
    Row: TableRow,
    Cell: TableCell,
    HeaderCell: TableHeaderCell,
    Foot: TableFoot,
});
Table.displayName = 'Table';
Table.Head.displayName = 'Table.Head';
Table.Body.displayName = 'Table.Body';
Table.Row.displayName = 'Table.Row';
Table.Cell.displayName = 'Table.Cell';
Table.HeaderCell.displayName = 'Table.HeaderCell';
Table.Foot.displayName = 'Table.Foot';

export { Table, TableBody, TableCell, TableFoot, TableHead, TableHeaderCell, TableRow };
