// @digdir/designsystemet-react/dist/cjs/components/Field/index.js
'use client';
'use strict';

var Field$1 = require('./Field.js');
var FieldAffix = require('./FieldAffix.js');
var FieldCounter = require('./FieldCounter.js');
var FieldDescription = require('./FieldDescription.js');

/**
 * Field component, used to wrap a form field.
 *
 * @example
 * <Field data-my-field>
 *   <Label>Kort beskrivelse</Label>
 *   <Field.Description>Beskrivelse</Field.Description>
 *   <Input />
 *   <ValidationMessage>Feilmelding</ValidationMessage>
 * </Field>
 */
const Field = Object.assign(Field$1.Field, {
    Description: FieldDescription.FieldDescription,
    Affixes: FieldAffix.FieldAffixes,
    Affix: FieldAffix.FieldAffix,
    Counter: FieldCounter.FieldCounter,
});
Field.Description.displayName = 'Field.Description';
Field.Affixes.displayName = 'Field.Affixes';
Field.Affix.displayName = 'Field.Affix';
Field.Counter.displayName = 'Field.Counter';

exports.FieldAffix = FieldAffix.FieldAffix;
exports.FieldAffixes = FieldAffix.FieldAffixes;
exports.FieldCounter = FieldCounter.FieldCounter;
exports.FieldDescription = FieldDescription.FieldDescription;
exports.Field = Field;
