// @digdir/designsystemet-react/dist/cjs/components/Dropdown/index.js
'use client';
'use strict';

var Dropdown$1 = require('./Dropdown.js');
var DropdownButton = require('./DropdownButton.js');
var DropdownHeading = require('./DropdownHeading.js');
var DropdownItem = require('./DropdownItem.js');
var DropdownList = require('./DropdownList.js');
var DropdownTrigger = require('./DropdownTrigger.js');
var DropdownTriggerContext = require('./DropdownTriggerContext.js');

/**
 * Dropdown component, used to display a list of options.
 *
 * @example
 * <Dropdown.TriggerContext>
 *   <Dropdown.Trigger>Dropdown trigger</Dropdown.Trigger>
 *   <Dropdown placement='bottom-end'>
 *     <Dropdown.Heading>Dropdown</Dropdown.Heading>
 *     <Dropdown.List>
 *       <Dropdown.Item>
 *         <Dropdown.Button>Option</Dropdown.Button>
 *       </Dropdown.Item>
 *     </Dropdown.List>
 *   </Dropdown>
 * </Dropdown.TriggerContext>
 */
const Dropdown = Object.assign(Dropdown$1.Dropdown, {
    TriggerContext: DropdownTriggerContext.DropdownTriggerContext,
    Heading: DropdownHeading.DropdownHeading,
    List: DropdownList.DropdownList,
    Item: DropdownItem.DropdownItem,
    Button: DropdownButton.DropdownButton,
    Trigger: DropdownTrigger.DropdownTrigger,
});
Dropdown.TriggerContext.displayName = 'Dropdown.TriggerContext';
Dropdown.List.displayName = 'Dropdown.List';
Dropdown.Heading.displayName = 'Dropdown.Heading';
Dropdown.Item.displayName = 'Dropdown.Item';
Dropdown.Button.displayName = 'Dropdown.Button';
Dropdown.Trigger.displayName = 'Dropdown.Trigger';

exports.DropdownButton = DropdownButton.DropdownButton;
exports.DropdownHeading = DropdownHeading.DropdownHeading;
exports.DropdownItem = DropdownItem.DropdownItem;
exports.DropdownList = DropdownList.DropdownList;
exports.DropdownTrigger = DropdownTrigger.DropdownTrigger;
exports.DropdownTriggerContext = DropdownTriggerContext.DropdownTriggerContext;
exports.Dropdown = Dropdown;
