// @digdir/designsystemet-react/dist/cjs/components/Suggestion/index.js
'use client';
'use strict';

var Suggestion = require('./Suggestion.js');
var SuggestionClear = require('./SuggestionClear.js');
var SuggestionEmpty = require('./SuggestionEmpty.js');
var SuggestionInput = require('./SuggestionInput.js');
var SuggestionList = require('./SuggestionList.js');
var SuggestionOption = require('./SuggestionOption.js');

/**
 * A component that provides a suggestion list for an input field.
 *
 * @example
 * <Suggestion>
 *   <Suggestion.Input />
 *   <Suggestion.Clear />
 *   <Suggestion.List>
 *     <Suggestion.Empty>Tomt</Suggestion.Empty>
 *     <Suggestion.Option value='Option 1'>Option 1</Suggestion.Option>
 *     <Suggestion.Option value='Option 2'>Option 2</Suggestion.Option>
 *   </Suggestion.List>
 * </Suggestion>
 */
const EXPERIMENTAL_Suggestion = Object.assign(Suggestion.Suggestion, {
    List: SuggestionList.SuggestionList,
    Input: SuggestionInput.SuggestionInput,
    Empty: SuggestionEmpty.SuggestionEmpty,
    Option: SuggestionOption.SuggestionOption,
    Clear: SuggestionClear.SuggestionClear,
});
EXPERIMENTAL_Suggestion.displayName = 'EXPERIMENTAL_Suggestion';
EXPERIMENTAL_Suggestion.List.displayName = 'EXPERIMENTAL_Suggestion.List';
EXPERIMENTAL_Suggestion.Input.displayName = 'EXPERIMENTAL_Suggestion.Input';
EXPERIMENTAL_Suggestion.Empty.displayName = 'EXPERIMENTAL_Suggestion.Empty';
EXPERIMENTAL_Suggestion.Option.displayName = 'EXPERIMENTAL_Suggestion.Option';
EXPERIMENTAL_Suggestion.Clear.displayName = 'EXPERIMENTAL_Suggestion.Clear';

exports.EXPERIMENTAL_SuggestionClear = SuggestionClear.SuggestionClear;
exports.EXPERIMENTAL_SuggestionEmpty = SuggestionEmpty.SuggestionEmpty;
exports.EXPERIMENTAL_SuggestionInput = SuggestionInput.SuggestionInput;
exports.EXPERIMENTAL_SuggestionList = SuggestionList.SuggestionList;
exports.EXPERIMENTAL_SuggestionOption = SuggestionOption.SuggestionOption;
exports.EXPERIMENTAL_Suggestion = EXPERIMENTAL_Suggestion;
