// @digdir/designsystemet-react/dist/cjs/components/MultiSuggestion/index.js
'use client';
'use strict';

var MultiSuggestion$1 = require('./MultiSuggestion.js');
var MultiSuggestionChips = require('./MultiSuggestionChips.js');
var MultiSuggestionClear = require('./MultiSuggestionClear.js');
var MultiSuggestionEmpty = require('./MultiSuggestionEmpty.js');
var MultiSuggestionInput = require('./MultiSuggestionInput.js');
var MultiSuggestionList = require('./MultiSuggestionList.js');
var MultiSuggestionOption = require('./MultiSuggestionOption.js');

const MultiSuggestion = Object.assign(MultiSuggestion$1.MultiSuggestion, {
    Input: MultiSuggestionInput.MultiSuggestionInput,
    List: MultiSuggestionList.MultiSuggestionList,
    Option: MultiSuggestionOption.MultiSuggestionOption,
    Chips: MultiSuggestionChips.MultiSuggestionChips,
    Empty: MultiSuggestionEmpty.MultiSuggestionEmpty,
    Clear: MultiSuggestionClear.MultiSuggestionClear,
});
MultiSuggestion.displayName = 'EXPERIMENTRAL_MultiSuggestion';
MultiSuggestion.Input.displayName = 'EXPERIMENTRAL_MultiSuggestion.Input';
MultiSuggestion.List.displayName = 'EXPERIMENTRAL_MultiSuggestion.List';
MultiSuggestion.Option.displayName = 'EXPERIMENTRAL_MultiSuggestion.Option';
MultiSuggestion.Chips.displayName = 'EXPERIMENTRAL_MultiSuggestion.Chips';
MultiSuggestion.Empty.displayName = 'EXPERIMENTRAL_MultiSuggestion.Empty';
MultiSuggestion.Clear.displayName = 'EXPERIMENTRAL_MultiSuggestion.Clear';

exports.EXPERIMENTAL_MultiSuggestionChips = MultiSuggestionChips.MultiSuggestionChips;
exports.EXPERIMENTAL_MultiSuggestionClear = MultiSuggestionClear.MultiSuggestionClear;
exports.EXPERIMENTAL_MultiSuggestionEmpty = MultiSuggestionEmpty.MultiSuggestionEmpty;
exports.EXPERIMENTAL_MultiSuggestionInput = MultiSuggestionInput.MultiSuggestionInput;
exports.EXPERIMENTAL_MultiSuggestionList = MultiSuggestionList.MultiSuggestionList;
exports.EXPERIMENTAL_MultiSuggestionOption = MultiSuggestionOption.MultiSuggestionOption;
exports.EXPERIMENTAL_MultiSuggestion = MultiSuggestion;
