// @digdir/designsystemet-react/dist/cjs/components/ErrorSummary/index.js
'use client';
'use strict';

var ErrorSummary$1 = require('./ErrorSummary.js');
var ErrorSummaryHeading = require('./ErrorSummaryHeading.js');
var ErrorSummaryItem = require('./ErrorSummaryItem.js');
var ErrorSummaryLink = require('./ErrorSummaryLink.js');
var ErrorSummaryList = require('./ErrorSummaryList.js');

/**
 * ErrorSummary component, used to display a list of errors.
 *
 * @example
 * <ErrorSummary>
 *   <ErrorSummary.Heading>Heading</ErrorSummary.Heading>
 *   <ErrorSummary.List>
 *     <ErrorSummary.Item>
 *       <ErrorSummary.Link href='#'>Error 1</ErrorSummary.Link>
 *     </ErrorSummary.Item>
 *     <ErrorSummary.Item>
 *       <ErrorSummary.Link href='#'>Error 2</ErrorSummary.Link>
 *     </ErrorSummary.Item>
 *   </ErrorSummary.List>
 * </ErrorSummary>
 */
const ErrorSummary = Object.assign(ErrorSummary$1.ErrorSummary, {
    Heading: ErrorSummaryHeading.ErrorSummaryHeading,
    Item: ErrorSummaryItem.ErrorSummaryItem,
    List: ErrorSummaryList.ErrorSummaryList,
    Link: ErrorSummaryLink.ErrorSummaryLink,
});
ErrorSummary.Item.displayName = 'ErrorSummary.Item';
ErrorSummary.Heading.displayName = 'ErrorSummary.Heading';
ErrorSummary.List.displayName = 'ErrorSummary.List';
ErrorSummary.Link.displayName = 'ErrorSummary.Link';

exports.ErrorSummaryHeading = ErrorSummaryHeading.ErrorSummaryHeading;
exports.ErrorSummaryItem = ErrorSummaryItem.ErrorSummaryItem;
exports.ErrorSummaryLink = ErrorSummaryLink.ErrorSummaryLink;
exports.ErrorSummaryList = ErrorSummaryList.ErrorSummaryList;
exports.ErrorSummary = ErrorSummary;
