// @digdir/designsystemet-react/dist/esm/components/MultiSuggestion/index.js
'use client';
import { MultiSuggestion as MultiSuggestion$1 } from './MultiSuggestion.js';
import { MultiSuggestionChips } from './MultiSuggestionChips.js';
import { MultiSuggestionClear } from './MultiSuggestionClear.js';
import { MultiSuggestionEmpty } from './MultiSuggestionEmpty.js';
import { MultiSuggestionInput } from './MultiSuggestionInput.js';
import { MultiSuggestionList } from './MultiSuggestionList.js';
import { MultiSuggestionOption } from './MultiSuggestionOption.js';

const MultiSuggestion = Object.assign(MultiSuggestion$1, {
    Input: MultiSuggestionInput,
    List: MultiSuggestionList,
    Option: MultiSuggestionOption,
    Chips: MultiSuggestionChips,
    Empty: MultiSuggestionEmpty,
    Clear: MultiSuggestionClear,
});
MultiSuggestion.displayName = 'EXPERIMENTRAL_MultiSuggestion';
MultiSuggestion.Input.displayName = 'EXPERIMENTRAL_MultiSuggestion.Input';
MultiSuggestion.List.displayName = 'EXPERIMENTRAL_MultiSuggestion.List';
MultiSuggestion.Option.displayName = 'EXPERIMENTRAL_MultiSuggestion.Option';
MultiSuggestion.Chips.displayName = 'EXPERIMENTRAL_MultiSuggestion.Chips';
MultiSuggestion.Empty.displayName = 'EXPERIMENTRAL_MultiSuggestion.Empty';
MultiSuggestion.Clear.displayName = 'EXPERIMENTRAL_MultiSuggestion.Clear';

export { MultiSuggestion as EXPERIMENTAL_MultiSuggestion, MultiSuggestionChips as EXPERIMENTAL_MultiSuggestionChips, MultiSuggestionClear as EXPERIMENTAL_MultiSuggestionClear, MultiSuggestionEmpty as EXPERIMENTAL_MultiSuggestionEmpty, MultiSuggestionInput as EXPERIMENTAL_MultiSuggestionInput, MultiSuggestionList as EXPERIMENTAL_MultiSuggestionList, MultiSuggestionOption as EXPERIMENTAL_MultiSuggestionOption };
