// @digdir/designsystemet-react/dist/cjs/components/Table/index.js
'use client';
'use strict';

var Table$1 = require('./Table.js');
var TableBody = require('./TableBody.js');
var TableCell = require('./TableCell.js');
var TableFoot = require('./TableFoot.js');
var TableHead = require('./TableHead.js');
var TableHeaderCell = require('./TableHeaderCell.js');
var TableRow = require('./TableRow.js');

/**
 * Table component, used to display tabular data. Renders a native HTML table element.
 *
 * @example
 * <Table>
 *   <Table.Head>
 *     <Table.Row>
 *       <Table.HeaderCell>Name</Table.HeaderCell>
 *       <Table.HeaderCell>Age</Table.HeaderCell>
 *     </Table.Row>
 *   </Table.Head>
 *   <Table.Body>
 *     <Table.Row>
 *       <Table.Cell>John</Table.Cell>
 *       <Table.Cell>25</Table.Cell>
 *     </Table.Row>
 *   </Table.Body>
 *   <Table.Foot>
 *     <Table.Row>
 *       <Table.Cell>Total</Table.Cell>
 *       <Table.Cell>2</Table.Cell>
 *     </Table.Row>
 *   </Table.Foot>
 * </Table>
 */
const Table = Object.assign(Table$1.Table, {
    Head: TableHead.TableHead,
    Body: TableBody.TableBody,
    Row: TableRow.TableRow,
    Cell: TableCell.TableCell,
    HeaderCell: TableHeaderCell.TableHeaderCell,
    Foot: TableFoot.TableFoot,
});
Table.displayName = 'Table';
Table.Head.displayName = 'Table.Head';
Table.Body.displayName = 'Table.Body';
Table.Row.displayName = 'Table.Row';
Table.Cell.displayName = 'Table.Cell';
Table.HeaderCell.displayName = 'Table.HeaderCell';
Table.Foot.displayName = 'Table.Foot';

exports.TableBody = TableBody.TableBody;
exports.TableCell = TableCell.TableCell;
exports.TableFoot = TableFoot.TableFoot;
exports.TableHead = TableHead.TableHead;
exports.TableHeaderCell = TableHeaderCell.TableHeaderCell;
exports.TableRow = TableRow.TableRow;
exports.Table = Table;
