// @digdir/designsystemet-react/dist/cjs/components/Tabs/index.js
'use client';
'use strict';

var Tabs$1 = require('./Tabs.js');
var TabsList = require('./TabsList.js');
var TabsPanel = require('./TabsPanel.js');
var TabsTab = require('./TabsTab.js');

/**
 * Display a group of tabs that can be toggled between.
 *
 * @example
 * <Tabs onChange={(value) => console.log(value)}>
 *   <Tabs.List>
 *     <Tabs.Tab value='1'>Tab 1</Tabs.Tab>
 *     <Tabs.Tab value='2'>Tab 2</Tabs.Tab>
 *     <Tabs.Tab value='3'>Tab 3</Tabs.Tab>
 *   </Tabs.List>
 *   <Tabs.Panel value='1'>content 1</Tabs.Panel>
 *   <Tabs.Panel value='2'>content 2</Tabs.Panel>
 *   <Tabs.Panel value='3'>content 3</Tabs.Panel>
 * </Tabs>
 */
const Tabs = Object.assign(Tabs$1.Tabs, {
    List: TabsList.TabsList,
    Tab: TabsTab.TabsTab,
    Panel: TabsPanel.TabsPanel,
});
Tabs.Tab.displayName = 'Tabs.Tab';
Tabs.List.displayName = 'Tabs.List';
Tabs.Panel.displayName = 'Tabs.Panel';

exports.TabsList = TabsList.TabsList;
exports.TabsPanel = TabsPanel.TabsPanel;
exports.TabsTab = TabsTab.TabsTab;
exports.Tabs = Tabs;
