// @digdir/designsystemet-react/dist/cjs/components/Breadcrumbs/index.js
'use client';
'use strict';

var Breadcrumbs$1 = require('./Breadcrumbs.js');
var BreadcrumbsItem = require('./BreadcrumbsItem.js');
var BreadcrumbsLink = require('./BreadcrumbsLink.js');
var BreadcrumbsList = require('./BreadcrumbsList.js');

/**
 * `Breadcrumbs` is a component that displays a list of breadcrumbs.
 *
 * @example
 * <Breadcrumbs aria-label='Du er her:' {...args}>
 *  <Breadcrumbs.Link href='#' aria-label='Tilbake til Niv책 1'>
 *    Niv책 1
 *  </Breadcrumbs.Link>
 *  <Breadcrumbs.List>
 *    <Breadcrumbs.Item>
 *      <Breadcrumbs.Link href='#'>Niv책 1</Breadcrumbs.Link>
 *    </Breadcrumbs.Item>
 *    <Breadcrumbs.Item>
 *      <Breadcrumbs.Link href='#'>Niv책 2</Breadcrumbs.Link>
 *    </Breadcrumbs.Item>
 *  </Breadcrumbs.List>
 * </Breadcrumbs>
 */
const Breadcrumbs = Object.assign(Breadcrumbs$1.Breadcrumbs, {
    List: BreadcrumbsList.BreadcrumbsList,
    Item: BreadcrumbsItem.BreadcrumbsItem,
    Link: BreadcrumbsLink.BreadcrumbsLink,
});
Breadcrumbs.List.displayName = 'Breadcrumbs.List';
Breadcrumbs.Item.displayName = 'Breadcrumbs.Item';
Breadcrumbs.Link.displayName = 'Breadcrumbs.Link';

exports.BreadcrumbsItem = BreadcrumbsItem.BreadcrumbsItem;
exports.BreadcrumbsLink = BreadcrumbsLink.BreadcrumbsLink;
exports.BreadcrumbsList = BreadcrumbsList.BreadcrumbsList;
exports.Breadcrumbs = Breadcrumbs;
