// @digdir/designsystemet-react/dist/cjs/components/Dialog/index.js
'use client';
'use strict';

var Dialog$1 = require('./Dialog.js');
var DialogBlock = require('./DialogBlock.js');
var DialogTrigger = require('./DialogTrigger.js');
var DialogTriggerContext = require('./DialogTriggerContext.js');

/**
 * Dialog component, used to display a Dialog dialog.
 *
 * @example
 * <Dialog.TriggerContext>
 *   <Dialog.Trigger>Open Dialog</Dialog.Trigger>
 *   <Dialog>
 *     <Dialog.Block>
 *       Content
 *     </Dialog.Block>
 *   </Dialog>
 * </Dialog.TriggerContext>
 */
const Dialog = Object.assign(Dialog$1.Dialog, {
    Block: DialogBlock.DialogBlock,
    TriggerContext: DialogTriggerContext.DialogTriggerContext,
    Trigger: DialogTrigger.DialogTrigger,
});
Dialog.Block.displayName = 'Dialog.Block';
Dialog.TriggerContext.displayName = 'Dialog.TriggerContext';
Dialog.Trigger.displayName = 'Dialog.Trigger';

exports.DialogBlock = DialogBlock.DialogBlock;
exports.DialogTrigger = DialogTrigger.DialogTrigger;
exports.DialogTriggerContext = DialogTriggerContext.DialogTriggerContext;
exports.Dialog = Dialog;
