// @digdir/designsystemet-react/dist/cjs/components/Search/index.js
'use client';
'use strict';

var Search$1 = require('./Search.js');
var SearchButton = require('./SearchButton.js');
var SearchClear = require('./SearchClear.js');
var SearchInput = require('./SearchInput.js');

/**
 * Search component, use to display different variations of a search input
 *
 * @example with button
 * <Search>
 *   <Search.Input aria-label='Søk' />
 *   <Search.Clear />
 *   <Search.Button />
 * </Search>
 *
 * @example with icon
 * <Search>
 *   <Search.Input aria-label='Søk' />
 *   <Search.Clear />
 * </Search>
 */
const Search = Object.assign(Search$1.Search, {
    Clear: SearchClear.SearchClear,
    Button: SearchButton.SearchButton,
    Input: SearchInput.SearchInput,
});
Search.Clear.displayName = 'Search.Clear';
Search.Button.displayName = 'Search.Button';
Search.Input.displayName = 'Search.Input';

exports.SearchButton = SearchButton.SearchButton;
exports.SearchClear = SearchClear.SearchClear;
exports.SearchInput = SearchInput.SearchInput;
exports.Search = Search;
