// @digdir/designsystemet-react/dist/esm/components/ErrorSummary/index.js
'use client';
import { ErrorSummary as ErrorSummary$1 } from './ErrorSummary.js';
import { ErrorSummaryHeading } from './ErrorSummaryHeading.js';
import { ErrorSummaryItem } from './ErrorSummaryItem.js';
import { ErrorSummaryLink } from './ErrorSummaryLink.js';
import { ErrorSummaryList } from './ErrorSummaryList.js';

/**
 * ErrorSummary component, used to display a list of errors.
 *
 * @example
 * <ErrorSummary>
 *   <ErrorSummary.Heading>Heading</ErrorSummary.Heading>
 *   <ErrorSummary.List>
 *     <ErrorSummary.Item>
 *       <ErrorSummary.Link href='#'>Error 1</ErrorSummary.Link>
 *     </ErrorSummary.Item>
 *     <ErrorSummary.Item>
 *       <ErrorSummary.Link href='#'>Error 2</ErrorSummary.Link>
 *     </ErrorSummary.Item>
 *   </ErrorSummary.List>
 * </ErrorSummary>
 */
const ErrorSummary = Object.assign(ErrorSummary$1, {
    Heading: ErrorSummaryHeading,
    Item: ErrorSummaryItem,
    List: ErrorSummaryList,
    Link: ErrorSummaryLink,
});
ErrorSummary.Item.displayName = 'ErrorSummary.Item';
ErrorSummary.Heading.displayName = 'ErrorSummary.Heading';
ErrorSummary.List.displayName = 'ErrorSummary.List';
ErrorSummary.Link.displayName = 'ErrorSummary.Link';

export { ErrorSummary, ErrorSummaryHeading, ErrorSummaryItem, ErrorSummaryLink, ErrorSummaryList };
