// @digdir/designsystemet-react/dist/cjs/components/Tooltip/Tooltip.js
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var dom = require('@floating-ui/dom');
var reactSlot = require('@radix-ui/react-slot');
var cl = require('clsx/lite');
var react = require('react');
var react$1 = require('@floating-ui/react');

/**
 * Tooltip component that displays a small piece of information when hovering or focusing on an element.
 *
 * @example
 * <Tooltip content='This is a tooltip'>
 *  <button>Hover me</button>
 * </Tooltip>
 *
 * @example
 * <Tooltip content='This is a tooltip'>
 *  Hover me
 * </Tooltip>
 */
const Tooltip = react.forwardRef(function Tooltip({ id, children, content, placement = 'top', open, className, ...rest }, ref) {
    const randomTooltipId = react.useId();
    const [internalOpen, setInternalOpen] = react.useState(false);
    const triggerRef = react.useRef(null);
    const tooltipRef = react.useRef(null);
    const mergedRefs = react$1.useMergeRefs([tooltipRef, ref]);
    const controlledOpen = open ?? internalOpen;
    const setOpen = () => {
        setInternalOpen(true);
    };
    const setClose = () => {
        setInternalOpen(false);
    };
    // Position with floating-ui
    react.useEffect(() => {
        const tooltip = tooltipRef.current;
        const trigger = triggerRef.current;
        tooltip?.togglePopover?.(controlledOpen);
        if (tooltip)
            tooltip.style.opacity = controlledOpen ? '1' : '0';
        if (tooltip && trigger && controlledOpen) {
            return dom.autoUpdate(trigger, tooltip, () => {
                dom.computePosition(trigger, tooltip, {
                    placement,
                    strategy: 'fixed',
                    middleware: [
                        dom.offset((data) => {
                            // get pseudo element arrow size
                            const styles = getComputedStyle(data.elements.floating, '::before');
                            return parseFloat(styles.height);
                        }),
                        dom.flip({
                            fallbackAxisSideDirection: 'start',
                        }),
                        dom.shift(),
                        arrowPseudoElement,
                    ],
                }).then(({ x, y }) => {
                    tooltip.style.translate = `${x}px ${y}px`;
                });
            });
        }
    }, [controlledOpen, placement]);
    /* Add listener for ESC to dismiss */
    react.useEffect(() => {
        const handleKeyDown = (event) => {
            if (event.key === 'Escape') {
                setInternalOpen(false);
            }
        };
        window.addEventListener('keydown', handleKeyDown);
        return () => {
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, []);
    /* If children is only a string, make a span */
    const ChildContainer = typeof children === 'string' ? 'span' : reactSlot.Slot;
    /* Make sure it is valid */
    if (typeof children !== 'string' && children.type === react.Fragment) {
        console.error('<Tooltip> children needs to be a single ReactElement that can receive a ref and not: <Fragment/> | <></>');
        return null;
    }
    const popoverProps = {
        [react.version.startsWith('19') ? 'popoverTarget' : 'popovertarget']: id ?? randomTooltipId,
        [react.version.startsWith('19')
            ? 'popoverTargetAction'
            : 'popovertargetaction']: 'show',
    };
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx(ChildContainer, { ref: triggerRef, ...popoverProps, onMouseEnter: setOpen, onMouseLeave: setClose, onFocus: setOpen, onBlur: setClose, children: children }), jsxRuntime.jsx("div", { ref: mergedRefs, role: 'tooltip', className: cl('ds-tooltip', className), id: id ?? randomTooltipId, 
                // @ts-ignore @types/react-dom does not understand popover yet
                popover: 'manual', ...rest, children: content })] }));
});
const arrowPseudoElement = {
    name: 'ArrowPseudoElement',
    fn(data) {
        const { elements, rects, placement } = data;
        let arrowX = `${Math.round(rects.reference.width / 2 + rects.reference.x - data.x)}px`;
        let arrowY = `${Math.round(rects.reference.height / 2 + rects.reference.y - data.y)}px`;
        switch (placement) {
            case 'top':
                arrowY = '100%';
                break;
            case 'right':
                arrowX = '0';
                break;
            case 'bottom':
                arrowY = '0';
                break;
            case 'left':
                arrowX = '100%';
                break;
        }
        elements.floating.style.setProperty('--dsc-tooltip-arrow-x', arrowX);
        elements.floating.style.setProperty('--dsc-tooltip-arrow-y', arrowY);
        return data;
    },
};

exports.Tooltip = Tooltip;
