// @digdir/designsystemet-react/dist/cjs/components/Combobox/useFloatingCombobox.js
'use client';
'use strict';

var react$1 = require('@floating-ui/react');
var react = require('react');
var reactDom = require('react-dom');
var ComboboxIdContext = require('./ComboboxIdContext.js');

const useFloatingCombobox = ({ listRef }) => {
    const [open, setOpen] = react.useState(false);
    const { activeIndex } = ComboboxIdContext.useComboboxId();
    const dispatch = ComboboxIdContext.useComboboxIdDispatch();
    // floating UI
    const { refs, floatingStyles, context } = react$1.useFloating({
        open,
        onOpenChange: (newOpen) => {
            if (!newOpen)
                dispatch?.({ type: 'SET_ACTIVE_INDEX', payload: 0 });
            reactDom.flushSync(() => {
                if (refs.floating.current && !newOpen) {
                    refs.floating.current.scrollTop = 0;
                }
                setTimeout(() => {
                    setOpen(newOpen);
                }, 1);
            });
        },
        whileElementsMounted: (reference, floating, update) => {
            react$1.autoUpdate(reference, floating, update);
            return () => {
                floating.scrollTop = 0;
            };
        },
        middleware: [
            react$1.flip({ padding: 10 }),
            react$1.size({
                apply({ rects, elements }) {
                    requestAnimationFrame(() => {
                        Object.assign(elements.floating.style, {
                            width: `${rects.reference.width}px`,
                            maxHeight: `200px`,
                        });
                    });
                },
            }),
            react$1.offset(10),
        ],
    });
    const role = react$1.useRole(context, { role: 'listbox' });
    const dismiss = react$1.useDismiss(context);
    const listNav = react$1.useListNavigation(context, {
        listRef,
        activeIndex,
        virtual: true,
        scrollItemIntoView: true,
        enabled: open,
        focusItemOnHover: true,
        onNavigate: (index) => {
            dispatch?.({ type: 'SET_ACTIVE_INDEX', payload: index || 0 });
        },
    });
    const { getReferenceProps, getFloatingProps, getItemProps } = react$1.useInteractions([role, dismiss, listNav]);
    return {
        open,
        setOpen,
        activeIndex,
        refs,
        floatingStyles,
        context,
        getReferenceProps,
        getFloatingProps,
        getItemProps,
    };
};

exports.useFloatingCombobox = useFloatingCombobox;
