// @digdir/designsystemet-react/dist/esm/components/Breadcrumbs/index.js
'use client';
import { Breadcrumbs as Breadcrumbs$1 } from './Breadcrumbs.js';
import { BreadcrumbsItem } from './BreadcrumbsItem.js';
import { BreadcrumbsLink } from './BreadcrumbsLink.js';
import { BreadcrumbsList } from './BreadcrumbsList.js';

/**
 * `Breadcrumbs` is a component that displays a list of breadcrumbs.
 *
 * @example
 * <Breadcrumbs aria-label='Du er her:' {...args}>
 *  <Breadcrumbs.Link href='#' aria-label='Tilbake til Niv책 1'>
 *    Niv책 1
 *  </Breadcrumbs.Link>
 *  <Breadcrumbs.List>
 *    <Breadcrumbs.Item>
 *      <Breadcrumbs.Link href='#'>Niv책 1</Breadcrumbs.Link>
 *    </Breadcrumbs.Item>
 *    <Breadcrumbs.Item>
 *      <Breadcrumbs.Link href='#'>Niv책 2</Breadcrumbs.Link>
 *    </Breadcrumbs.Item>
 *  </Breadcrumbs.List>
 * </Breadcrumbs>
 */
const Breadcrumbs = Object.assign(Breadcrumbs$1, {
    List: BreadcrumbsList,
    Item: BreadcrumbsItem,
    Link: BreadcrumbsLink,
});
Breadcrumbs.List.displayName = 'Breadcrumbs.List';
Breadcrumbs.Item.displayName = 'Breadcrumbs.Item';
Breadcrumbs.Link.displayName = 'Breadcrumbs.Link';

export { Breadcrumbs, BreadcrumbsItem, BreadcrumbsLink, BreadcrumbsList };
